core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,222 @@
  * necessary.
  */
 class DelayedStream implements ClientStream {
-
-  // set to non null once both listener and realStream are valid.  After this point it is safe
-  // to call methods on startedRealStream.  Note: this may be true even after the delayed stream is
-  // cancelled.  This should be okay.
-  private volatile ClientStream startedRealStream;
-  @GuardedBy("this")
-  private String authority;
-  @GuardedBy("this")
+  /** {@code true} once realStream is valid and all pending calls have been drained. */
+  private volatile boolean passThrough;
+  /**
+   * Non-{@code null} iff start has been called. Used to assert methods are called in appropriate
+   * order, but also used if an error occurrs before {@code realStream} is set.
+   */
   private ClientStreamListener listener;
-  @GuardedBy("this")
   private ClientStream realStream;
+  /** The stream was cancelled before {@code realStream} was set. */
   @GuardedBy("this")
   private Status error;
-
-  @GuardedBy("this")
-  private final List<PendingMessage> pendingMessages = new LinkedList<PendingMessage>();
-  private boolean messageCompressionEnabled;
-  @GuardedBy("this")
-  private boolean pendingHalfClose;
-  @GuardedBy("this")
-  private int pendingFlowControlRequests;
-  @GuardedBy("this")
-  private boolean pendingFlush;
   @GuardedBy("this")
-  private Compressor compressor;
-  @GuardedBy("this")
-  private Decompressor decompressor;
-
-  static final class PendingMessage {
-    final InputStream message;
-    final boolean shouldBeCompressed;
+  private List<Runnable> pendingCalls = new ArrayList<Runnable>();
 
-    public PendingMessage(InputStream message, boolean shouldBeCompressed) {
-      this.message = message;
-      this.shouldBeCompressed = shouldBeCompressed;
+  /**
+   * Transfers all pending and future requests and mutations to the given stream.
+   *
+   * <p>No-op if either this method or {@link #cancel} have already been called.
+   */
+  final void setStream(ClientStream stream) {
+    synchronized (this) {
+      if (error != null || realStream != null) {
+        return;
+      }
+      realStream = checkNotNull(stream, "stream");
     }
-  }
 
-  @Override
-  public synchronized void setAuthority(String authority) {
-    checkState(listener == null, "must be called before start");
-    checkNotNull(authority, "authority");
-    if (realStream == null) {
-      this.authority = authority;
-    } else {
-      realStream.setAuthority(authority);
+    List<Runnable> toRun = new ArrayList<Runnable>();
+    while (true) {
```

Comment:
This part is pretty tricky! It's worth a comment to explain why the loop is necessary. IIUC, it's because 1) runnable has to be run outside the lock; 2) passThrough can only be set to true when pendingCalls is empty, and this must be done inside the lock. So while runnable are running outside the lock, new runnables may be added to pendingCalls.